### PR TITLE
First draft at auto reloading policies

### PR DESCRIPTION
**Pull request by aaronjeline**
Issue #34 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: tinytodo/src/policy_store.rs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+use crate::context::{AppQuery, AppQueryKind};
+
+#[tracing::instrument]
+fn watch_function(result: std::result::Result<Event, Error>) {
+    trace!("Got notify event");
+    match result {
+        Ok(ev) => match ev.kind {
+            EventKind::Access(AccessKind::Close(AccessMode::Write)) => todo!(),
+            EventKind::Remove(RemoveKind::File) => {
+                info!("Policy set file updated!")
```
Comment (mwhicks1) [inline tinytodo/src/policy_store.rs:38]: But this doesn't seem to be doing anything?

File: tinytodo/src/context.rs [hunk 2]
```diff
@@ -266,12 +277,32 @@ impl AppContext {
...
                 };
-                msg.sender.send(r).unwrap();
+                if let Err(e) = msg.sender.send(r) {
+                    trace!("Failed send response: {:?}", e);
+                }
             }
         }
     }
 
+    #[tracing::instrument]
```
Comment (mwhicks1) [inline tinytodo/src/context.rs:290]: Remove all debugging traces etc. before merging.

Comment (mwhicks1) [reply tinytodo/src/context.rs:290]: (here and elsewhere in this file)

File: tinytodo/src/policy_store.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                Ok(event) => {
+                    trace!("Event: {:?}", event);
+                    match event.kind {
+                        EventKind::Access(AccessKind::Close(AccessMode::Write)) => {
+                            let (send, _recv) = tokio::sync::oneshot::channel();
+                            let kind = AppQueryKind::UpdatePolicySet;
+                            let q = AppQuery::new(kind, send);
+                            tx.blocking_send(q).expect("Failed to send");
+                        }
+                        EventKind::Remove(_) => {
```
Comment (mwhicks1) [inline tinytodo/src/policy_store.rs:71]: What is the `Remove` event? Removing the watch? Or removing the file?

File: tinytodo/src/policy_store.rs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        )
+        .expect("Failed to create watcher");
+
+        let mut s = Self {
+            watcher,
+            path: PathBuf::from(path),
+        };
+
+        s.set_watch();
+
```
Comment (mwhicks1) [inline tinytodo/src/policy_store.rs:96]: Format properly? I'm suspicious of the extra spaces.

**Review by mwhicks1 — COMMENTED** [review]
I'm not clear how this works, so asked some questions. Remove the debugging info?

File: tinytodo/src/context.rs (hunk 2 above)
Comment (aaronjeline) [reply tinytodo/src/context.rs:290]: These are not debugging traces, these are for `tracing`, the logging library we use.

File: tinytodo/src/policy_store.rs (hunk 1 above)
Comment (aaronjeline) [reply tinytodo/src/policy_store.rs:38]: Ack you're right. Had some git versioning issues and think I pushed the wrong version.

File: tinytodo/src/policy_store.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        &self.path
+    }
+
+    #[tracing::instrument]
+    pub fn new(tx: Sender<AppQuery>, path: &Path) -> Self {
+        let config = Config::default().with_poll_interval(Duration::from_secs(1));
+        let watcher = RecommendedWatcher::new(
+            move |res: Result<Event, Error>| match res {
+                Ok(event) => {
+                    trace!("Event: {:?}", event);
```
Comment (mwhicks1) [inline tinytodo/src/policy_store.rs:48]: I don't think I want to see this event.

**Review by mwhicks1 — APPROVED** [review]
I don't understand the precise logic, but I am guessing this watcher library is checking for file changes, then you get an event and read in the new file. If so: good! Just a small comment to drop a traced event.

File: tinytodo/src/policy_store.rs (hunk 5 above)
Comment (aaronjeline) [reply tinytodo/src/policy_store.rs:48]: You only see if you have the log level set to `trace`, one of the higher log levels. (Ie: setting it to `info` won't show this)

Comment (aaronjeline) [reply tinytodo/src/policy_store.rs:48]: And yes, `watcher` abstracts over the all the file watching APIs across different operating systems

**Review by shaobo-he-aws — APPROVED** [review]

**Pull request merged**